### PR TITLE
Add tags for datasets

### LINKED ISSUE #5829: Organization of datasets with tags
## Detailed Description
Similar to how tags can be assigned to annotations, tag-supported could be added to facilitate the management of datasets. Predominantly, this would allow:

- adding arbitrary meta data to datasets
- tag-based search and filtering

## Context

* [ ] Specific to long-running jobs (set `jobsEnabled=true` in `application.conf`)
* [ ] Specific to webKnossos.org (set `isDemoInstance=true` in `application.conf`)

### PR DESCRIPTION
**Pull request by fm3**
Adding a tag system to Datasets, analogous to what Annotations already have.

### URL of deployed dev instance (used for testing):
- https://datasettags.webknossos.xyz

### Backend Changes
- Dataset json now contains additional field `tags`  (list of string)
- Dataset update route now expects `tags` as well (it being missing will be interpreted as empty list), changes tags in database

### Steps to test:
- Open the dataset table. There should be a new row for tags.
- Try adding some tags for a few datasets.
- Clicking on one of the tags should add the tag to the search and hide all datasets that do not have this tag.
- The clicked tag should appear on the top left of the dataset table.
- Clicking on close on the list of tags the datasets are filtered for, should remove that tag from the filtering
- Reloading the page should preserve the filtered tags.
- Adding or removing a tag from a dataset, then waiting a second for the backend call and then reloading the whole page, should show the dataset with the added/removed tag. 
- Reloading a dataset should also not change the displayed tags of a dataset.

### Issues:
- fixes #5829 

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [x] Ready for review

### CODE REVIEW

**Comment by fm3** [comment]
@MichaelBuessemeyer The backend part should be done. I could only test it in a limited way so far, as there is no frontend yet. If you encounter problems, feel free to ping me :)

File: tools/postgres/schema.sql [hunk 1]
```diff
@@ -98,6 +98,7 @@ CREATE TABLE webknossos.dataSets(
   logoUrl VARCHAR(2048),
   sortingKey TIMESTAMPTZ NOT NULL,
   details JSONB,
+  tags VARCHAR(256)[] NOT NULL DEFAULT '{}',
```
Comment (jstriebel) [inline tools/postgres/schema.sql:101]: We could also use the `json` or `jsonb` type (seems like it's available from pg 9.6). It's also possible to create indices on jsonb columns and query them efficiently (e.g. when searching/selecting by tag):

https://www.postgresql.org/docs/9.4/datatype-json.html#JSON-INDEXING

**Review by jstriebel — COMMENTED** [review]
Backend mostly LGTM, please see my two comments @fm3:

File: tools/postgres/schema.sql (hunk 1 above)
Comment (jstriebel) [reply tools/postgres/schema.sql:101]: But also totally fine as-is for now I guess.

File: app/models/binary/DataSetService.scala [hunk 2]
```diff
@@ -374,6 +374,7 @@ class DataSetService @Inject()(organizationDAO: OrganizationDAO,
         "isUnreported" -> Json.toJson(isUnreported(dataSet)),
         "isForeign" -> dataStore.isForeign,
         "jobsEnabled" -> jobsEnabled,
+        "tags" -> dataSet.tags
```
[the remark this answers is not in the document]
Comment (fm3) [reply app/models/binary/DataSetService.scala:377]: So far we have not considered additional fields a breaking change. This would increase the number of new versions by a lot.
I agree that we should unify which routes are versioned and what constitutes a new version. I created a separate issue for this at https://github.com/scalableminds/webknossos/issues/5840

**Review by philippotto — CHANGES_REQUESTED** [review]
The front-end code looks great :+1: However, I think that I've found a potential race condition (see my comment).

Other than that, I have some UI suggestions:
- turn the cursor into a pointer when hovering the tags (via CSS)
- show a tooltip which explains the tag interactions (e.g., "Click to only show datasets with this tag", "Add a new tag", "Remove this tag from this dataset")

Would be great if you could generalize these UI improvements so that the tags for the explorative annotations also profit from these :)

File: conf/evolutions/reversions/077-add-dataset-tags.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+START TRANSACTION;
+
+DROP VIEW webknossos.dataSets_;
+
+ALTER TABLE webknossos.dataSets DROP COLUMN tags;
+
+CREATE VIEW webknossos.dataSets_ AS SELECT * FROM webknossos.dataSets WHERE NOT isDeleted;
+
+UPDATE webknossos.releaseInformation SET schemaVersion = 76;
```
Comment (philippotto) [inline conf/evolutions/reversions/077-add-dataset-tags.sql:9]: the version number in the file name and in this line also need to be incremented by one (I think, this is why the CI fails)

Comment (fm3) [reply conf/evolutions/reversions/077-add-dataset-tags.sql:9]: Correct, since on master the version has now been incremented to 77, the verison in schema.sql needs to be 78 here now. Also, both evolution files should be renamed to 078*, the links in migrations.md updated, as well as the numbers incremented as you said (this was done for the forward evolution already above). The CI does not check the reversion, though. It fails because schema.sql Line 24 was not incremented to 78.

File: frontend/javascripts/dashboard/dataset/dataset_cache_provider.js [hunk 4]
```diff
@@ -155,9 +168,41 @@ export default function DatasetCacheProvider({ children }: { children: Node }) {
     }
   }
 
+  async function updateCachedDataset(dataset: APIDataset) {
+    if (isLoading) return;
+    setIsLoading(true);
```
[the remark this answers is not in the document]
Comment (MichaelBuessemeyer) [reply frontend/javascripts/dashboard/dataset/dataset_cache_provider.js]: Thanks for spotting this. I implemented a queueing like behaviour for all updates on the same dataset. This queuing ensures that each previous request is finished before the current request starts. Therefore it is ensured that the backend always gets the updates in the correct order. Thus no update call that (by network magic) was faster at the server gets lost because a slower update request that is much slower overwrites its changes afterwards.
Here is a screenshot of a few dataset changes made on a single dataset in a row by changing the tags of a single dataset. The connection is throttled very strongly to test this corner case. The graph on the right illustrates that one call only starts after the previous updates finished.
image

File: frontend/javascripts/dashboard/dataset/dataset_cache_provider.js [hunk 5]
```diff
@@ -155,9 +171,48 @@ export default function DatasetCacheProvider({ children }: { children: Node }) {
...
+        return dataset;
+      } else {
+        return currentDataset;
+      }
+    });
+    setDatasets(updatedDatasets);
+    datasetCache.set(updatedDatasets);
+    try {
+      const previousDatasetUpdatePromise = pendingDatasetUpdates[dataset.name];
+      const newDatasetUpdatePromise = Promise.resolve(previousDatasetUpdatePromise).then(async () =>
```
Comment (philippotto) [inline frontend/javascripts/dashboard/dataset/dataset_cache_provider.js:191]: While the behavior seems to be correct, I stumbled upon this, because it confused me that no null-check is done and I personally only use `Promise.resolve` to create a fulfilled Promise with a dedicated value. I would either add a comment which explains why the code works or rewrite it like this:

```suggestion
      const previousDatasetUpdatePromise = pendingDatasetUpdates[dataset.name] || Promise.resolve();
      const newDatasetUpdatePromise = previousDatasetUpdatePromise.then(() =>
```

Also note that I removed the `async` keyword as it wasn't necessary.

File: CHANGELOG.unreleased.md [hunk 6]
```diff
@@ -11,7 +11,14 @@ For upgrade instructions, please check the [migration guide](MIGRATIONS.released
 [Commits](https://github.com/scalableminds/webknossos/compare/21.11.0...HEAD)
 
 ### Added
--
+- Enhanced the volume fill tool to so that it operates beyond the dimensions of the current viewport. Additionally, the fill tool can also be changed to perform in 3D instead of 2D. [#5733](https://github.com/scalableminds/webknossos/pull/5733)
```
Comment (philippotto) [inline CHANGELOG.unreleased.md:14]: this is probably a merge error?

File: frontend/javascripts/dashboard/dataset/dataset_cache_provider.js (hunk 5 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/dashboard/dataset/dataset_cache_provider.js:191]: I am fine with your version :)

I just learned about this behaviour of `Promise.resolve` while reading on how to create a "promise-queue" :)

File: CHANGELOG.unreleased.md (hunk 6 above)
Comment (MichaelBuessemeyer) [reply CHANGELOG.unreleased.md]: Yeah i just found them and tried to fix it :/

**Review by philippotto — APPROVED** [review]
Great, looks merge-ready to me :)

**Comment by MichaelBuessemeyer** [comment]
Is it correct, that I had to adjust the version of the `tools/postgres/schema.sql` manually? 

If yes, hopefully the CI now works 😅

File: frontend/javascripts/dashboard/dataset/dataset_cache_provider.js [hunk 7]
```diff
@@ -44,12 +56,16 @@ export const DatasetCacheContext = createContext<Context>({
...
-  updateDataset: async () => {},
+  reloadDataset: async () => {},
+  updateCachedDataset: async () => {},
 });
 
 export default function DatasetCacheProvider({ children }: { children: Node }) {
   const [datasets, setDatasets] = useState(datasetCache.get());
   const [isLoading, setIsLoading] = useState(false);
+  const [pendingDatasetUpdates, setPendingDatasetUpdates] = useState<{
+    [string]: Array<APIDataset>,
```
Comment (philippotto) [inline frontend/javascripts/dashboard/dataset/dataset_cache_provider.js:67]: This should probably be

```suggestion
    [string]: Promise<APIDataset>,
```

to fix the flow error.
